### DISCUSSION TITLE
Path to custom ES index template unknown

### DISCUSSION THREAD

**Discussion by ben-sec** [Security-Onion-Solutions/securityonion#6173] [Unsupported Versions] [answered]
Hello!

I followed [these instructions](https://github.com/Security-Onion-Solutions/securityonion/discussions/6004#discussioncomment-1581585) to declare a new custom ES index template, but now I'm wondering what the correct path to this template is, that I have to use in my logstash output config file? E.g. the logstash configuration for syslog outputs sets template => "/templates/so-syslog-template.json", but if I set template => "/templates/my-custom-template.json" I get an error in my logstash.log (File does not exist or cannot be opened). So what is the correct path to my template file, that is located in /opt/so/saltstack/local/salt/elasticsearch/templates/custom that I have to use in my logstash output config file? Thanks in advance!

Cheers, Ben

**Comment by weslambert** [comment, marked as answer]
Hi @ben-sec ,

The path defined in Logstash the output is relative to the bind mount for the Docker container (`/templates/` in the container)(`/opt/so/conf/elasticsearch/templates` on the filesystem):

https://github.com/Security-Onion-Solutions/securityonion/blob/master/salt/logstash/init.sls#L164

So, when you are referencing a template in there, you would need to refer to the path from the container's perspective.

(The container sees `/templates/my-template.json`, while it is `/opt/so/conf/elasticsearch/templates/my-template.json` on disk)

However, in order for the template to even end up inside of `/opt/so/conf/elasticsearch/templates`, it would need to be put inside of `/opt/so/saltstack/local/salt/elasticsearch/templates/so|custom` to override the default or append to the default.

The default templates (`/opt/so/saltstack/default|local/salt/elasticseach/tempaltes/so`) are already referenced in the Elasticsearch pillar to be copied to `/opt/so/conf/elasticsearch/templates` when the Salt state is applied. To add additional templates, you would need to specify the relative path to the directory that the container understands.

The template path defined in the elasticsearch pillar is relative to `/opt/so/saltstack/default/salt/elasticsearch/templates/so|custom`
and determines what actually gets copied into the resultant `/opt/so/conf/elasticsearch/templates` directory from which Logstash will load the template specified in the output config, and the `so-elasticsearch-templatesload` script will load the templates into Elasticsearch.

You don't necessarily have to provide the template in the Logstash output, but it helps to add another mechanism to ensure it's loaded whenever the index is created.

Assuming your custom template is in `/opt/so/saltstack/local/salt/elasticsearch/templates/custom`, you'll need to ensure it's loaded by specifying the path in the Elasticsearch pillar file in `/opt/so/saltstack/local/pillar/elasticsearch/search.sls`, as:

`- custom/my-template.json`

**Comment by ben-sec** [comment]
Hi Wes,

OK, I skipped the template in logstash, but now I run in an aweful long error (starting with message=>"Unable to configure plugins: Illegal character in path at index 0: {{ ..." and I'm wondering if the host statement in the following code (copied from configuration files found on the search node) is actually working on a manager node:

```
output {
  if "beacons" in [tags]{
        elasticsearch {
        pipeline => "rita-beacons"
        hosts => "{{ ES }}"
        index => "rita-beacons"
        template_name => "rita-beacons"
        #template => "/templates/rita-beacons-template.json"
        template_overwrite => true
        ssl => true
        ssl_certificate_verification => false
        }
     }
  }
```

By the way: Do I have to worry that I can't find my custom template in the Kibana index management yet (since no data has been imported yet)?

Again, thanks for helping me out. I hope I will be able to import Rita results in Elasticsearch ... 

Cheers, Ben

**Reply by weslambert** [reply]
It looks like you are now remarking the template, but still have configuration referencing insertion of a template.  Have you tried removing ALL of the template config?

Mappings won't be visible for the index until data is actually indexed into Elasticsearch.

**Reply by ben-sec** [reply]
Now I'm getting confused: Are you saying that I don't need custom index templates and the steps as explained [here](https://github.com/Security-Onion-Solutions/securityonion/discussions/6004#discussioncomment-1581585) at all? Ich will give it a try tomorrow. However, is the way I reference to my Elasticsearch host above correct on a manager node or? Thanks, Wes!

**Comment by ben-sec** [comment]
OK, now I nearly got it: I had to understand the whole jinja-template mechanics in order to get things working in my logstash output section. Now I have at least the data in Elasticsearch, but my custom template is not working yet. In my logstash.log it says:

`Failed to install template {:message=>"Got response code '400' contacting Elasticsearch at URL 'https://manager-ip:9200/_template/rita-beacons'"`

Something seems to be not at the right place yet ... Ideas?

**Reply by weslambert** [reply]
Have you double-checked the date of the error message in the Logstash log?  Have you confirmed there is no lingering Logstash config that is trying to apply the template? Also, if you are still trying to use the Logstash output to apply the template, have you confirmed the template is visible inside the Logstash Docker container?

What is the output of the following?

`sudo so-elasticsearch-templates-list | grep rita`

**Comment by ben-sec** [comment]
Hi Wes, I stopped the template troubleshooting, since (at least for me) it involves to much magic to get things working and I have already archieved my main goal to import RITA beacon activity in ES for visualizations without a custom ES template. Works for me in this case ... However, did you guys ever think about integrating RITA in a future SO release? I think the guys at BHIS did a great job there! Thanks again!

**Reply by weslambert** [reply]
Hi Ben,

I agree that the folks at BHIS have done an excellent job with RITA!

We've unofficially integrated with RITA in the past, but given the way that the data is handled, it sometimes makes it a little more difficult to manage when you have multiple sensors, given that you either have to run RITA on each node, or ship logs off to a completely separate node.  As far as beacon data ingestion, it shouldn't be too difficult to support consuming that information back into SO (as you have done).  It may be something we look into in the future, but it's not on our roadmap at this time.

**Reply by presianbg** [reply]
@ben-sec , are you willing to share what and how you did to integrate RITA with SO ? 

Cheers,
PY

**Reply by ben-sec** [reply]
Sure! Be aware that the following is done in a distributed SO environment!

1. Install RITA by following the official instructions [here](https://github.com/activecm/rita)  (Note: I was only able to install RITA on my Manager node, not the Forward node (installation failed without any error messages), where the zeek-logs are located, which you will need).
2. I created a custom logstash pipeline by adding the following logstash configuration in /opt/so/saltstack/local/salt/logstash/pipelines/config/custom/beacons-import.conf.jinja

```
{%- if grains['role'] == 'so-eval' -%}
{%- set ES = salt['pillar.get']('manager:mainip', '') -%}
{%- else %}
{%- set ES = salt['pillar.get']('elasticsearch:mainip', '') -%}
{%- endif %}

input {
	file {
	      path => "/var/log/logstash/rita/pipeline/beacons.csv"
	      tags => "beacons"
	      mode => "tail"
	      start_position => "beginning"	      
	     }
}

filter { if "beacons" in [tags]
	 {
	 csv {
	     separator => ","
	     skip_header => true
	     columns => ["score","source_ip","destination_ip","connections","avg_bytes","intvl_range","size_range","top_intvl","top_size","top_intvl_count","top_size_count","intvl_skew","size_skew","intvl_dispersion","size_dispersion","total_bytes"]
	     }

	 mutate {add_field => {"destination_url" => "%{destination_ip}"}}
	 mutate {add_field => {"source_url" => "%{source_ip}"}}
	 dns {
	     action => "replace"
	     nameserver => ["IP-of-your-local-name-server"]
	     reverse => [ "destination_url","source_url" ]
	     hit_cache_size => 4096
    	     hit_cache_ttl => 900
	     failed_cache_size => 512
    	     failed_cache_ttl => 900
	     }

	geoip {source=>"destination_ip"}

        }
}

output {
if "beacons" in [tags]{
      elasticsearch {
      pipeline => "rita-beacons"
      hosts => "{{ ES }}"
      index => "rita-beacons"
      ssl => true
      ssl_certificate_verification => false
      }
   }
}
```

3. Declare this new pipeline in `/opt/so/saltstack/local/pillar/minions/<MINIONID>_manager.sls` on your Manager node
```
logstash:
     pipelines:
       manager:
         config:
           - so/0009_input_beats.conf
           - so/0010_input_hhbeats.conf
           - so/9999_output_redis.conf.jinja
           - custom/beacons-import.conf.jinja
```

4. Declare a new empty ES ingest node pipeline called "rita-beacons" in Kibana Stack Management.
5. Copy zeek logs from your Forward Node to your Master (e.g. yesterdays logs as a daily cron job):
```
/usr/bin/sshpass -f "/home/so/pw.txt" scp -r so@IP-of-your-forward-node:/nsm/zeek/logs/`date -d"-1day" +"%Y-%m-%d"` /opt/so/log/logstash/rita/import/zeek-`date -d"-1day" +"%Y-%m-%d"`
```
In my example /home/so/pw.txt must contain the password of your so account (or you do it manually without sshpass ...).

6. Run RITA on the Manager node:
```
/usr/local/bin/rita import --rolling --numchunks 1 /opt/so/log/logstash/rita/import/zeek-`date -d"-1day" +"%Y-%m-%d"` yesterday
```
7. Export Beacon results to CSV
```
/usr/local/bin/rita show-beacons yesterday > /opt/so/log/logstash/rita/pipeline/beacons.csv
```
8. Restart Logstash on your Manager node (so-logstash-restart) and verify that Logstash has started without problems with so-status.
9. Now you should find an index called "rita-beacons" in Elasticsearch that contains the beacon signals of yesterday. Whenever you overwrite beacons.csv, Logstash will import the new file automatically. Therefore I put steps 5, 6 and 7 in a daily cron job (+ one additional job to delete the zeek-logs on the Manager Node after the RITA import).
10. Create an index pattern (e.g. rita-*) and start to visualize your beacon data in Kibana!

I hope I didn't miss a thing ...

Cheers, Ben

**Reply by presianbg** [reply]
Hi, great writeup! 
I will test it ASAP. 

I'm running RITA on all forward nodes without any issues, but haven't ingested the data in ELK yet, because had not idea how to implement the pipeline. 
Its very odd that the RITA is failing to build / install on your forward nodes. 

The downside of your solution is that you have to copy zeek logs from X sensors and probably stressing the manager node when processing with RITA. This seems not very scalable  solution, but as Wes said - completely separate node could be used just for RITA.

**Reply by ben-sec** [reply]
Thanks! Did you already install SO 2.3.90? It looks like my custom pipelines broke Logstash on the Search node (please refer to [this discussion](https://github.com/Security-Onion-Solutions/securityonion/discussions/6341#discussion-3702655)).

**Reply by ben-sec** [reply]
@presianbg The configuration of step 3 has to go to `/opt/so/saltstack/local/pillar/minions/<MINIONID>_manager.sls`, otherwise you break Logstash on your search node in SO 2.3.90.

**Reply by ghost** [reply]
> Sure! Be aware that the following is done in a distributed SO environment!
> 
> 1. Install RITA by following the official instructions [here](https://github.com/activecm/rita)  (Note: I was only able to install RITA on my Manager node, not the Forward node (installation failed without any error messages), where the zeek-logs are located, which you will need).
> 2. I created a custom logstash pipeline by adding the following logstash configuration in /opt/so/saltstack/local/salt/logstash/pipelines/config/custom/beacons-import.conf.jinja
> 
> ```
> {%- if grains['role'] == 'so-eval' -%}
> {%- set ES = salt['pillar.get']('manager:mainip', '') -%}
> {%- else %}
> {%- set ES = salt['pillar.get']('elasticsearch:mainip', '') -%}
> {%- endif %}
> 
> input {
> 	file {
> 	      path => "/var/log/logstash/rita/pipeline/beacons.csv"
> 	      tags => "beacons"
> 	      mode => "tail"
> 	      start_position => "beginning"	      
> 	     }
> }
> 
> filter { if "beacons" in [tags]
> 	 {
> 	 csv {
> 	     separator => ","
> 	     skip_header => true
> 	     columns => ["score","source_ip","destination_ip","connections","avg_bytes","intvl_range","size_range","top_intvl","top_size","top_intvl_count","top_size_count","intvl_skew","size_skew","intvl_dispersion","size_dispersion","total_bytes"]
> 	     }
> 
> 	 mutate {add_field => {"destination_url" => "%{destination_ip}"}}
> 	 mutate {add_field => {"source_url" => "%{source_ip}"}}
> 	 dns {
> 	     action => "replace"
> 	     nameserver => ["IP-of-your-local-name-server"]
> 	     reverse => [ "destination_url","source_url" ]
> 	     hit_cache_size => 4096
>     	     hit_cache_ttl => 900
> 	     failed_cache_size => 512
>     	     failed_cache_ttl => 900
> 	     }
> 
> 	geoip {source=>"destination_ip"}
> 
>         }
> }
> 
> output {
> if "beacons" in [tags]{
>       elasticsearch {
>       pipeline => "rita-beacons"
>       hosts => "{{ ES }}"
>       index => "rita-beacons"
>       ssl => true
>       ssl_certificate_verification => false
>       }
>    }
> }
> ```
> 
> 3. Declare this new pipeline in /opt/so/saltstack/local/pillar/global.sls on your Manager node
> 
> ```
> logstash:
>      pipelines:
>        manager:
>          config:
>            - so/0009_input_beats.conf
>            - so/0010_input_hhbeats.conf
>            - so/9999_output_redis.conf.jinja
>            - custom/beacons-import.conf.jinja
> ```
> 
> 4. Declare a new empty ES ingest node pipeline called "rita-beacons" in Kibana Stack Management.
> 5. Copy zeek logs from your Forward Node to your Master (e.g. yesterdays logs as a daily cron job):
> 
> ```
> /usr/bin/sshpass -f "/home/so/pw.txt" scp -r so@IP-of-your-forward-node:/nsm/zeek/logs/`date -d"-1day" +"%Y-%m-%d"` /opt/so/log/logstash/rita/import/zeek-`date -d"-1day" +"%Y-%m-%d"`
> ```
> 
> In my example /home/so/pw.txt must contain the password of your so account (or you do it manually without sshpass ...).
> 
> 6. Run RITA on the Manager node:
> 
> ```
> /usr/local/bin/rita import --rolling --numchunks 1 /opt/so/log/logstash/rita/import/zeek-`date -d"-1day" +"%Y-%m-%d"` yesterday
> ```
> 
> 7. Export Beacon results to CSV
> 
> ```
> /usr/local/bin/rita show-beacons yesterday > /opt/so/log/logstash/rita/pipeline/beacons.csv
> ```
> 
> 8. Restart Logstash on your Manager node (so-logstash-restart) and verify that Logstash has started without problems with so-status.
> 9. Now you should find an index called "rita-beacons" in Elasticsearch that contains the beacon signals of yesterday. Whenever you overwrite beacons.csv, Logstash will import the new file automatically. Therefore I put steps 5, 6 and 7 in a daily cron job (+ one additional job to delete the zeek-logs on the Manager Node after the RITA import).
> 10. Create an index pattern (e.g. rita-*) and start to visualize your beacon data in Kibana!
> 
> I hope I didn't miss a thing ...
> 
> Cheers, Ben

Hi @ben-sec,
I currently have a so-standalone machine and trying to integrate Vulnwhisperer with SO. The configuration's of Vulnwhisperer looks pretty similar to that of RITA. When I run the logstash vulnwhisperer config, I am not able to find anything on elasticsearch. The very same configuration is working for me in an Ubuntu machine. And I don't have any entries of logstash on global.sls file. Can you let me know if my setup itself is wrong? And any help regarding loading custom logstash configurations just like how you have here would be helpful. 

Thanks

**Reply by ben-sec** [reply]
Hi @Y3ll0wF1ash, to be honest, I'm not very familiar with standalone environments and how the directory structure looks like on a single machine for Logstash pipelines. But you mentioned that you find no entries  regarding logstash in global.sls. That's key to success (please refer to my step 3 above and the official SO documentation [here](https://docs.securityonion.net/en/2.3/logstash.html?highlight=logstash#adding-new-logs)). In my case it turned out, that you have to add the custom pipeline config to `/opt/so/saltstack/local/pillar/minions/<MINIONID>_manager.sls`.
